addons/sale_loyalty/models/sale_order.py

```diff
@@ -0,0 +1,964 @@
+        elif reward.discount_mode == 'percent':
+            max_discount = min(max_discount, discountable * (reward.discount / 100))
+        # Discount per taxes
+        reward_code = _generate_random_reward_code()
+        point_cost = reward.required_points if not reward.clear_wallet else self._get_real_points_for_coupon(coupon)
+        if reward.discount_mode == 'per_point' and not reward.clear_wallet:
+            # Calculate the actual point cost if the cost is per point
+            converted_discount = self.currency_id._convert(min(max_discount, discountable), reward.currency_id, self.company_id, fields.Date.today())
+            point_cost = converted_discount / reward.discount
+        # Gift cards and eWallets are considered gift cards and should not have any taxes
+        if reward.program_id.program_type in ('ewallet', 'gift_card'):
+            return [{
+                'name': reward.description,
+                'product_id': reward.discount_line_product_id.id,
+                'price_unit': -min(max_discount, discountable),
+                'product_uom_qty': 1.0,
+                'product_uom': reward.discount_line_product_id.uom_id.id,
+                'reward_id': reward.id,
+                'coupon_id': coupon.id,
+                'points_cost': point_cost,
+                'reward_identifier_code': reward_code,
+            }]
+        discount_factor = min(1, (max_discount / discountable)) if discountable else 1
+        mapped_taxes = {tax: self.fiscal_position_id.map_tax(tax) for tax in discountable_per_tax}
+        reward_dict = {tax: {
+            'name': _(
+                'Discount: %(desc)s%(tax_str)s',
+                desc=reward.description,
+                tax_str=_(' - On product with the following taxes: %(taxes)s', taxes=", ".join(mapped_taxes[tax].mapped('name'))),
+            ),
+            'product_id': reward.discount_line_product_id.id,
+            'price_unit': -(price * discount_factor),
+            'product_uom_qty': 1.0,
+            'product_uom': reward.discount_line_product_id.uom_id.id,
+            'reward_id': reward.id,
+            'coupon_id': coupon.id,
+            'points_cost': 0,
+            'reward_identifier_code': reward_code,
+            'tax_id': [(Command.CLEAR, 0, 0)] + [(Command.LINK, tax.id, False) for tax in mapped_taxes[tax]]
+        } for tax, price in discountable_per_tax.items()}
+        # We only assign the point cost to one line to avoid counting the cost multiple times
+        if reward_dict:
+            reward_dict[next(iter(reward_dict))]['points_cost'] = point_cost
+        # Returning .values() directly does not return a subscribable list
+        return list(reward_dict.values())
+
+    def _get_program_domain(self):
+        """
+        Returns the base domain that all programs have to comply to.
+        """
+        self.ensure_one()
+        return [('active', '=', True),
```

Comment:
Same for these ones no ? The ORM is supposed to add ``('active', '=', True)`` automatically.

Reply:
I guess I just wanted to have them set explicitely, as it is more clear when jumping into the code but indeed it is redundant.

Reply:
Actually since it is used in some filtered_domain, it is necessary to filter out applied programs on active.